### ISSUE TITLE
 If you search a non existing emoji and press enter, then after deleting the search term it won't fetch emojis again - reported by @Tushu17 

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8287]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
 1. Go to Chat > Emoji
 2. Search an non-existing emoji
 3. Press Enter
 4. Empty search box

## Expected Result:
It should show emojis after clearing the search term.

## Actual Result:
It shows 'No result found'

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- Desktop App

**Version Number:** 1.1.44-4
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/159847584-b358c6e3-ccd4-4f42-a26a-9f0afa0abf5b.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Tushu17
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1647102086741009

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by sobitneupane** [comment]
**Proposal:**
Addition of following line of code will solve the issue:
https://github.com/Expensify/App/blob/6e9d429ad47044d71cb028fb14d5defc2de0af8a/src/components/EmojiPicker/EmojiPickerMenu/index.js#L152-L159

```diff
            if (this.searchInput && !this.searchInput.isFocused()) {
                this.setState({selectTextOnFocus: false});
                this.searchInput.value = '';
                this.searchInput.focus();
+               this.filterEmojis(this.searchInput.value);
                
                // Re-enable selection on the searchInput
                this.setState({selectTextOnFocus: true});
            }

```

**Comment by jliexpensify** [comment]
Posted

Internal - https://www.upwork.com/ab/applicants/1508261111247921152/job-details
External - https://www.upwork.com/jobs/~01b2bf9b46c621dad9

**Comment by Puneet-here** [comment]
### Proposal 
Need to remove this line 
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154

it will also solve one more problem. As of now when we press enter and type something the previous search term get's cleared instead it should stay and user should be able to modify it. 

**Before :-** 

https://user-images.githubusercontent.com/101883770/160362109-784e6c2c-c471-469b-b762-e713083b411d.mp4

Problem: if the user mistyped and presses enter then the user will just correct the typo but the search term is getting cleared so the user has to type the whole search term again which is not a good ux.

**After removing the line 154:-** 

https://user-images.githubusercontent.com/101883770/160362103-54fac1da-7f60-44bf-9fcf-47dd345beec7.mp4

**Comment by Santhosh-Sellavel** [comment]
@Puneet-here  
Sound goods. But on pressing enter focus should not be last right?

**Comment by Puneet-here** [comment]
> @Puneet-here Sound goods. But on pressing enter focus should not be lost. Can you update your proposal?

we can pass `onBlur={() => this.searchInput.focus()}` in composer

**Comment by Santhosh-Sellavel** [comment]
Will review this ASAP (today or by tomorrow).

**Comment by Santhosh-Sellavel** [comment]
https://github.com/Expensify/App/issues/8287#issuecomment-1086899013
That does not make any sense. @Puneet-here Because it will allow me to focus away from the search box.

Did you find out why the focus is lost first?

**Comment by jliexpensify** [comment]
Bump @Puneet-here - can you review @Santhosh-Sellavel 's comment?

Also, re-assigning as I will be OOO for 2 weeks.

**Comment by JmillsExpensify** [comment]
Jumping in for @jliexpensify while he's out. 🙌

**Comment by Santhosh-Sellavel** [comment]
Let's double this @JmillsExpensify 

cc: @Luke9389

**Comment by Santhosh-Sellavel** [comment]
Proposals lack why this issue occurs (root cause), no follow-ups

**Comment by JmillsExpensify** [comment]
Project budget doubled to $500.

**Comment by sobitneupane** [comment]
**Proposal:**
Reason: Following lines of code change searchInput value to empty.
https://github.com/Expensify/App/blob/6e9d429ad47044d71cb028fb14d5defc2de0af8a/src/components/EmojiPicker/EmojiPickerMenu/index.js#L152-L159

But It does not filterEmojis. So, I proposed following solution.
> **Proposal:** Addition of following line of code will solve the issue:
> 
> https://github.com/Expensify/App/blob/6e9d429ad47044d71cb028fb14d5defc2de0af8a/src/components/EmojiPicker/EmojiPickerMenu/index.js#L152-L159
> 
> ```diff
>             if (this.searchInput && !this.searchInput.isFocused()) {
>                 this.setState({selectTextOnFocus: false});
>                 this.searchInput.value = '';
>                 this.searchInput.focus();
> +               this.filterEmojis(this.searchInput.value);
>                 
>                 // Re-enable selection on the searchInput
>                 this.setState({selectTextOnFocus: true});
>             }
> ```

What is the expected output?

**Comment by Puneet-here** [comment]
> [#8287 (comment)](https://github.com/Expensify/App/issues/8287#issuecomment-1086899013) That does not make any sense. @Puneet-here Because it will not allow me to focus away from the search box.

I don't think there's any problem in that. Can you tell why the user will try to take the focus away? I couldn't think of any practical reason. There will be only one case where the user will try to take the focus away from the search: when user wants to write something in the compose field but tapping anywhere on the screen closes the emoji picker so there's no problem in that.

I don't think there's any other simple way of keeping the text cursor in the search box, can you suggest any alternative way?

> Did you find out why the focus is lost first?

As far as I know that's how inputs work, pressing enter on a field will take the focus away.

**Comment by Santhosh-Sellavel** [comment]
@Puneet-here 
>  Can you tell why the user will try to take the focus away?
Because we have arrow navigations to select emoji, so adding `onBlur={() => this.searchInput.focus()}` would break it.

**Comment by Santhosh-Sellavel** [comment]
@Luke9389 

Root Cause,
https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154-L155

In the above line while `this.searchInput.value = ''` or replacing it with "this.searchInput.clear()" should have triggered the `onChangeText={this.filterEmojis}` but it doesn’t. So the emoji's not updated properly.
https://github.com/Expensify/App/blob/ad637d3dd7dbb03cfdc6ebcb9196944c4f6165bb/src/components/EmojiPicker/EmojiPickerMenu/index.js#L438-L442

In @sobitneupane proposal [here](https://github.com/Expensify/App/issues/8287#issuecomment-1103570835) by manually calling it `this.filterEmojis(this.searchInput.value);` the issue is resolved, but this looks like a workaround.

But the root cause with `Composer` would still exist. 

We can
1. wait for proposals to solve it at the root (which we prefer)
or
2. Just go ahead with @sobitneupane which looks good to me as well.

Let me know your thoughts thanks!

**Comment by Puneet-here** [comment]
> Because we have arrow navigations to select emoji, so adding `onBlur={() => this.searchInput.focus()}` would break it.

Oh I never saw this feature. Hmm.. I don't think there's any other way to achieve that.

> In @sobitneupane proposal https://github.com/Expensify/App/issues/8287#issuecomment-1103570835 by manually calling it this.filterEmojis(this.searchInput.value); the issue is resolved, but this looks like a workaround.

I don't think it's a good way to fix it also it doesn't fix one thing.
I am proposing that we remove the line 154 which is the real cause of this and my way isn't a workaround
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154

If we go with Sobit's proposal the search term will be cleared every time user presses backspace instead it should stay and user should be able to modify it.

This is what we will get from my approach:

https://user-images.githubusercontent.com/101883770/160362103-54fac1da-7f60-44bf-9fcf-47dd345beec7.mp4


Is there any reason for not going with my proposal?

@Luke9389 can you share your thoughts it would be great.

**Comment by Santhosh-Sellavel** [comment]
> > In @sobitneupane proposal [#8287 (comment)](https://github.com/Expensify/App/issues/8287#issuecomment-1103570835) by manually calling it this.filterEmojis(this.searchInput.value); the issue is resolved, but this looks like a workaround.
> 
> I don't think it's a good way to fix it also it doesn't fix one thing.

@Puneet-here 
1. What one thing doesn't fix?

> the search term will be cleared every time
Seems It was designed that way.

The reasons I didn’t choose your proposal, was adding `onBlur={() => this.searchInput.focus()}`, without adding that it is a probable option to fix this issue.

cc: @Luke9389

**Comment by Puneet-here** [comment]
>The reasons I didn’t choose your proposal, was adding onBlur={() => this.searchInput.focus()}, without adding that it is a probable option to fix this issue.

My original proposal was to just remove line 154 which is the real cause of this issue, we can just leave the onBlur.

https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154

and I don't think search term getting cleared is correct.
We should not clear the whole search term as the user will try to correct his search term and if we clear it user will have to type the whole search term again.

**Comment by Santhosh-Sellavel** [comment]
Waiting for your thoughts @Luke9389

My comment -> https://github.com/Expensify/App/issues/8287#issuecomment-1107608662

**Comment by JmillsExpensify** [comment]
@Luke9389 I know you're prioritizing offline at the moment, though please chime in when you have a free cycle. Thank you!

**Comment by Luke9389** [comment]
Hey everyone, thanks for waiting.
I was OOO last week. 

> replacing it with "this.searchInput.clear()" should have triggered the onChangeText={this.filterEmojis} but it doesn’t. 

Do you understand why this is the case? I think we should try to address the root cause if possible.

**Comment by Luke9389** [comment]
Also I just tried to reproduce this locally on the master dev branch and this seems to be fixed 🤔

**Comment by Santhosh-Sellavel** [comment]
Sorry I missed to comment back, I was able to reproduce this one still.

**Comment by Puneet-here** [comment]
> Also I just tried to reproduce this locally on the master dev branch and this seems to be fixed 🤔

Hi @Luke9389, It's reproducible can you try these steps:
1. Search a non-existing emoji
2. Press Enter
3. Press backspace without focusing again on the search field

**Comment by eVoloshchak** [comment]
**Proposal 1**
Change this line
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154

to 
```javascript 
this.searchInput.setNativeProps({value: ' '}); 
```

**Proposal 2**
If we need to fix this in the Composer itself, so the issue doesn't happen in other components, but this feels kinda hacky.
In Composer's [componentDidUpdate](https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/Composer/index.js#L151) method add
```javascript 
if (!this.props.value && this.prevValue !== this.textInput.value) {
    this.props.onChangeText(this.textInput.value);
}

this.prevValue = this.textInput.value;
```
Both solutions yield the same result:

https://user-images.githubusercontent.com/9059945/167296551-6dcc6db5-2b84-4ee7-9da8-f409958dc54d.mp4

**Comment by Santhosh-Sellavel** [comment]
@Luke9389 
@eVoloshchak [Proposal](https://github.com/Expensify/App/issues/8287#issuecomment-1120411714) looks good to me.

We are only using `setNativeProps` API on `textInputRef` to update the text on the other views that use Composer.
So Proposal 1 itself looks good to me!

🎀👀🎀 C+ reviewed.

**Comment by Puneet-here** [comment]
I don't think we need to use `this.searchInput.setNativeProps({value: ' '}); ` because there's no use of it.

if we do that this will be the result ( notice how the whole search term gets cleared on pressing backspace )

https://user-images.githubusercontent.com/101883770/167710820-4a7aa6bf-4f0d-4988-aa4a-4c8ee3fa9c51.mov

**What's wrong with it?**
As far as I can think of the user will see this issue when he makes a typo and presses enter and it's normal that the user will modify the search term to fix the typo. Now the problem is that pressing backspace will clear everything so I don't think it's a good solution.
That's why I am proposing that we just remove line 154 as it's the root cause of this issue.

https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/components/EmojiPicker/EmojiPickerMenu/index.js#L154

After removing line 154 (notice that pressing the backspace only doesn't clear the search term completely)

https://user-images.githubusercontent.com/101883770/167711931-a3e13cd3-f6da-4895-a6ca-d1879cccf278.mov

cc: @Luke9389

**Comment by Luke9389** [comment]
Hey folks. Thanks for your patience on this one. I'm working on a big project internally right now.

**Comment by JmillsExpensify** [comment]
@Luke9389 Thanks for the herculean effort over the past week. I *think* you'll be able to come up for a breathe of air and get back to this one this week? Jump in when you're able, thanks!

**Comment by Luke9389** [comment]
Hey folks. I know I've done this once already, but are you sure this issue is still happening? I've tried to reproduce and can't. Please test it out, take a screen recording, and tell me exactly what you're doing. 

I tried [this](https://github.com/Expensify/App/issues/8287#issuecomment-1118715900) and it worked fine.

**Comment by eVoloshchak** [comment]
@Luke9389 

https://user-images.githubusercontent.com/9059945/168970936-d14b5354-5d70-47fd-a093-7d265e29af4a.mp4

1. Search for a non-existing emoji
2. Press Enter
3. Press Backspace

**Comment by Luke9389** [comment]
Got it, OK. Thanks

**Comment by Luke9389** [comment]
I think I actually like @Puneet-here's idea to just take that line out. 
It's strange that a single backspace press would clear the whole line.

**Comment by NehaSantani** [comment]
Hey what is the status of this issue?
Is it still open?
I can work on it!

**Comment by Santhosh-Sellavel** [comment]
@Luke9389,

Especially when searching emojis we're gonna type  2- 5 characters not more than 10 - 15 characters. Moreover, we are not clear on everything on each backspace, only after the user press enter.  Considering we are not searching for anything important like a contact etc. Clearing doesn’t sound strange at all to me.

**Comment by Puneet-here** [comment]
I don't think we should clear the search term even if the search term isn't 10-15 character long. Why we wanna clear it when we should be keeping the search term there as the inputs are expected to work like that?

I don't think these are fair points for why we should clear the search term. We can solve this issue and improve something then why should we go with a workaround to solve only the issue?

**Comment by Santhosh-Sellavel** [comment]
@Luke9389 

I had started a slack 🧵 regarding the confusion here about clearing or keeping the invalid search input and modifying it. 
Most of the team prefer modifying the text consensus reach in the poll [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1653336003554859?thread_ts=1653335197.924749&cid=C01GTK53T8Q)

So Let's go with @Puneet-here proposal here!
🎀 👀 🎀 C+ reviewed

Thanks!

**Comment by Luke9389** [comment]
Ok, go ahead and start your PR @Puneet-here 👍

**Comment by Puneet-here** [comment]
I think this is happening with all the recent PRs

**Comment by JmillsExpensify** [comment]
> Hi @JmillsExpensify, the previous job has expired can you please open a new one.

Here you go! https://www.upwork.com/jobs/~019a6d1271f2483032 cc @Santhosh-Sellavel

**Comment by Puneet-here** [comment]
> Here you go! https://www.upwork.com/jobs/~019a6d1271f2483032 cc @Santhosh-Sellavel

Thank you @JmillsExpensify, I have applied to the job.

**Comment by JmillsExpensify** [comment]
@Puneet-here @Santhosh-Sellavel offers sent. @Tushu17 Can you apply as well for reporting? Thank you!

**Comment by JmillsExpensify** [comment]
Almost through the regression hold. Will circle back here in a couple of days for payment.

**Comment by Santhosh-Sellavel** [comment]
Lol, We didn’t assign this to @Puneet-here. 
@JmillsExpensify 
Please assign it to them before closing this out.
So far no regressions. Payment is due tomorrow.

**Comment by JmillsExpensify** [comment]
Thanks! Payments issued via Upwork to all contributors. Have a great weekend.
